### PR TITLE
Customized TestCase for pyiron

### PR DESCRIPTION
**Pull request by sudarsan-surendralal**
This PR includes the following changes

- Creating a generic `PyironTestCase` for all unittesting within pyiron which includes generic functions that can be used for all test modules
- Inheriting all test modules from this particular module
- Automatic testing of the docstrings of the `DataContainer` module

### CODE REVIEW

File: pyiron_base/_tests.py [hunk 1]
```diff
@@ -47,35 +72,10 @@ def tearDownClass(cls):
             pass
 
 
-class TestWithCleanProject(TestWithProject, ABC):
+class TestWithCleanProject(TestWithProject):
```
Comment (liamhuber) [inline pyiron_base/_tests.py]:
```suggestion
class TestWithCleanProject(TestWithProject, ABC):
```

We still expect this to be a parent for individual testing suites.

File: pyiron_base/_tests.py [hunk 2]
```diff
@@ -27,7 +27,32 @@
...
+
+        Output capturing adapted from https://stackoverflow.com/a/22434594/12332968
+        """
+        with StringIO() as buf, redirect_stdout(buf):
+            result = doctest.testmod(self.docstring_module)
+            output = buf.getvalue()
+        self.failIf(result.failed > 0, msg=output)
+
+
+class TestWithProject(PyironTestCase):
```
Comment (liamhuber) [inline pyiron_base/_tests.py]:
```suggestion
class TestWithProject(PyironTestCase, ABC):
```

We still expect this to be a parent for individual testing suites.

**Review by liamhuber — APPROVED** [review]
Please keep `ABC` for the all the parent test cases. Otherwise looks good!

**Comment by sudarsan-surendralal** [comment]
> 
> 
> Please keep `ABC` for the all the parent test cases. Otherwise looks good!

But `PyironTestCase` is derived from `ABC`. So isn't that sufficient?

**Comment by liamhuber** [comment]
> > Please keep `ABC` for the all the parent test cases. Otherwise looks good!
> 
> But `PyironTestCase` is derived from `ABC`. So isn't that sufficient?

Nope. I thought this was because inheriting from `ABC` makes you uninstantiable, but in fact that's only the case if you *also* have `abstractmethod`s.

What remains is that `ABC` indicates to the developer that this class is not intended to ever be instantiated, but only to act as a parent class for other classes. This is definitely the case for what we're doing here since these only serve as parents for actual test cases, but I will downgrade my complaint to a suggestion.

**Comment by sudarsan-surendralal** [comment]
> What remains is that `ABC` indicates to the developer that this class is not intended to ever be instantiated, but only to act as a parent class for other classes

Thanks, I get it now. Applied your changes.

**Pull request merged**